File: frontend-web/src/utils/viewPreloader.js
/**
 * Préchargement intelligent des vues clés
 * Précharge les composants probables avant navigation
 */
import { prefetchManager } from './prefetch';

class ViewPreloader {
  constructor() {
    this.preloadedViews = new Set();
    this.preloadQueue = [];
  }

  /**
   * Précharger une vue
   */
  async preloadView(viewName, preloadFn) {
    if (this.preloadedViews.has(viewName)) {
      return;
    }

    try {
      await preloadFn();
      this.preloadedViews.add(viewName);
    } catch (error) {
      console.warn(`Failed to preload view ${viewName}:`, error);
    }
  }

  /**
   * Précharger les vues probables après connexion
   */
  preloadKeyViews(userId) {
    // Précharger le dashboard
    // Note: apiClient a déjà baseURL avec /api, donc on utilise juste le chemin sans /api
    this.preloadView('dashboard', () => {
      return prefetchManager.prefetch('/dashboard', `dashboard:${userId}`);
    });

    // Précharger les fichiers récents
    this.preloadView('recentFiles', () => {
      return prefetchManager.prefetch('/files?limit=10', `recent:${userId}`);
    });
  }

  /**
   * Précharger au hover d'un lien
   */
  preloadOnHover(element, viewName, preloadFn) {
    let timeout;
    
    element.addEventListener('mouseenter', () => {
      timeout = setTimeout(() => {
        this.preloadView(viewName, preloadFn);
      }, 100); // Précharger après 100ms de hover
    });

    element.addEventListener('mouseleave', () => {
      if (timeout) {
        clearTimeout(timeout);
      }
    });
  }
}

export const viewPreloader = new ViewPreloader();

File: mobile-app/lib/utils/view_preloader.dart
/// Préchargement intelligent des vues clés
/// Précharge les données probables avant navigation
import 'advanced_cache.dart';
import '../services/api_service.dart';
import 'dart:async';

class ViewPreloader {
  static final ViewPreloader _instance = ViewPreloader._internal();
  factory ViewPreloader() => _instance;
  ViewPreloader._internal();

  final Set<String> _preloadedViews = {};
  final AdvancedCache _cache = AdvancedCache();

  /// Précharger une vue
  Future<void> preloadView(String viewName, Future<dynamic> Function() preloadFn) async {
    if (_preloadedViews.contains(viewName)) {
      return;
    }

    try {
      await preloadFn();
      _preloadedViews.add(viewName);
    } catch (e) {
      // Ignorer erreurs silencieusement
    }
  }

  /// Précharger les vues probables après connexion
  Future<void> preloadKeyViews(String userId) async {
    final apiService = ApiService();

    // Précharger le dashboard (pas besoin de /api car baseUrl l'inclut déjà)
    await preloadView('dashboard', () async {
      final response = await apiService.get('/dashboard');
      await _cache.set('dashboard:$userId', response.data, priority: CachePriority.high);
      return response.data;
    });

    // Précharger les fichiers récents (pas besoin de /api car baseUrl l'inclut déjà)
    await preloadView('recentFiles', () async {
      final response = await apiService.get('/files', queryParameters: {'limit': 10});
      await _cache.set('recent:$userId', response.data, priority: CachePriority.high);
      return response.data;
    });
  }

  /// Précharger au hover d'un élément (pour web)
  void preloadOnHover(String viewName, Future<dynamic> Function() preloadFn) {
    // Pour mobile, on peut précharger au tap prolongé ou au scroll
    // Implémentation simplifiée
  }
}

